bson/src/main/org/bson/RawBsonValueHelper.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson;
+
+import org.bson.codecs.BsonValueCodecProvider;
+import org.bson.codecs.DecoderContext;
+import org.bson.codecs.configuration.CodecRegistry;
+
+import static org.bson.codecs.BsonValueCodecProvider.getClassForBsonType;
+import static org.bson.codecs.configuration.CodecRegistries.fromProviders;
+
+final class RawBsonValueHelper {
+    private static final CodecRegistry REGISTRY = fromProviders(new BsonValueCodecProvider());
+
+    static BsonValue decode(final byte[] bytes, final BsonBinaryReader bsonReader) {
+        if (bsonReader.getCurrentBsonType() == BsonType.DOCUMENT || bsonReader.getCurrentBsonType() == BsonType.ARRAY) {
+            int position = bsonReader.getBsonInput().getPosition();
+            int size = bsonReader.getBsonInput().readInt32();
```

Comment:
Given this method is only used internally and we control the use of it, is it an assumption that the BsonBinaryReader hasn't yet read the size?

Also, I'm not sure what you envisage when you say by adding a getCurrentDocumentSize() method to BsonBinaryReader would abstract away the assumption . Do we need a new public method to add to the API?

Reply:
It's an assumption in that it's not part of any contract documented in BsonBinaryReader as to when the size is read from its BsonInput.  I'd certainly be wary of any external application making that assumption.

The fact that it's an internal method in the same package that's making this assumption makes it more palatable, and if we in the future changes BsonBinaryReader so that it is no longer true, tests included in this PR would start to fail.  So I'm inclined to not worry about this.

Yes, I envisaged getCurrentDocumentSize()  as a new public method on BsonBinaryReader.  That would allow an external application to write a method like this safely without relying on implementation details of BsonBinaryReader's use of BsonInput.  If we added it now, it just future proofs this method against such a change.